// @firestitch/example/esm2015/app/services/fs-example.service.js
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { of, forkJoin } from 'rxjs';
import { catchError, map } from 'rxjs/operators';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common/http";
export class FsExampleService {
    constructor(http) {
        this.http = http;
    }
    getFileContents(path, name) {
        const fileCodeRequests = this._paths(path, name).reduce((acc, file) => {
            const request = this.http.get(file.path, { responseType: 'text' }).pipe(map(code => ({ type: file.type, name: file.name, code })), catchError((error) => of(null)));
            acc.push(request);
            return acc;
        }, []);
        return forkJoin(...fileCodeRequests).pipe(map((files) => files.filter((file) => !!file)));
    }
    _paths(path, name) {
        const componentPath = path || name;
        return [
            {
                type: 'ts',
                name: `${name}.component.ts`,
                path: `${FsExampleService.EXAMPLE_PATH_PREFIX}/${componentPath}/${name}.component.ts`,
            },
            {
                type: 'html',
                name: `${name}.component.html`,
                path: `${FsExampleService.EXAMPLE_PATH_PREFIX}/${componentPath}/${name}.component.html`,
            },
            {
                type: 'scss',
                name: `${name}.component.scss`,
                path: `${FsExampleService.EXAMPLE_PATH_PREFIX}/${componentPath}/${name}.component.scss`,
            },
        ];
    }
}
FsExampleService.EXAMPLE_PATH_PREFIX = 'assets/components';
FsExampleService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsExampleService, deps: [{ token: i1.HttpClient }], target: i0.ɵɵFactoryTarget.Injectable });
FsExampleService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsExampleService });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsExampleService, decorators: [{
            type: Injectable
        }], ctorParameters: function () { return [{ type: i1.HttpClient }]; } });
