// @firestitch/label/bundles/firestitch-label.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/common')) :
    typeof define === 'function' && define.amd ? define('@firestitch/label', ['exports', '@angular/core', '@angular/common'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global.firestitch = global.firestitch || {}, global.firestitch.label = {}), global.ng.core, global.ng.common));
})(this, (function (exports, i0, common) { 'use strict';

    function _interopNamespace(e) {
        if (e && e.__esModule) return e;
        var n = Object.create(null);
        if (e) {
            Object.keys(e).forEach(function (k) {
                if (k !== 'default') {
                    var d = Object.getOwnPropertyDescriptor(e, k);
                    Object.defineProperty(n, k, d.get ? d : {
                        enumerable: true,
                        get: function () { return e[k]; }
                    });
                }
            });
        }
        n["default"] = e;
        return Object.freeze(n);
    }

    var i0__namespace = /*#__PURE__*/_interopNamespace(i0);

    var FsLabelComponent = /** @class */ (function () {
        function FsLabelComponent() {
        }
        return FsLabelComponent;
    }());
    FsLabelComponent.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0__namespace, type: FsLabelComponent, deps: [], target: i0__namespace.ɵɵFactoryTarget.Component });
    FsLabelComponent.ɵcmp = i0__namespace.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.16", type: FsLabelComponent, selector: "fs-label", ngImport: i0__namespace, template: "<ng-content></ng-content>", styles: [":host{transform-origin:0 0;transition:none;display:flex;color:#0000008a;font-size:75%;transform:translateY(-.15em) perspective(100px);min-height:12px}\n"], changeDetection: i0__namespace.ChangeDetectionStrategy.OnPush });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0__namespace, type: FsLabelComponent, decorators: [{
                type: i0.Component,
                args: [{
                        selector: 'fs-label',
                        templateUrl: 'label.component.html',
                        styleUrls: ['label.component.scss'],
                        changeDetection: i0.ChangeDetectionStrategy.OnPush,
                    }]
            }], ctorParameters: function () { return []; } });

    var FsLabelFieldComponent = /** @class */ (function () {
        function FsLabelFieldComponent() {
            this.bottomMargin = true;
            this.topMargin = true;
            this.labelMargin = true;
        }
        Object.defineProperty(FsLabelFieldComponent.prototype, "setBottomMargin", {
            set: function (value) {
                this.bottomMargin = value;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(FsLabelFieldComponent.prototype, "setTopMargin", {
            set: function (value) {
                this.topMargin = value;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(FsLabelFieldComponent.prototype, "setLabelMargin", {
            set: function (value) {
                this.labelMargin = value;
            },
            enumerable: false,
            configurable: true
        });
        return FsLabelFieldComponent;
    }());
    FsLabelFieldComponent.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0__namespace, type: FsLabelFieldComponent, deps: [], target: i0__namespace.ɵɵFactoryTarget.Component });
    FsLabelFieldComponent.ɵcmp = i0__namespace.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.16", type: FsLabelFieldComponent, selector: "fs-label-field", inputs: { setBottomMargin: ["bottomMargin", "setBottomMargin"], setTopMargin: ["topMargin", "setTopMargin"], setLabelMargin: ["labelMargin", "setLabelMargin"] }, host: { properties: { "class.field-bottom-margin": "this.bottomMargin", "class.field-top-margin": "this.topMargin", "class.field-label-margin": "this.labelMargin" } }, ngImport: i0__namespace, template: "<div class=\"field-wrap\">\n  <div class=\"field-label\"><ng-content select=\"fs-label\"></ng-content></div>\n\n  <div class=\"field-content\">\n    <ng-content></ng-content>\n  </div>\n</div>\n\n<div class=\"field-message\">\n  <ng-content select=\"fs-label-message\"></ng-content>\n</div>\n", styles: [":host{display:block;position:relative}:host.field-label-margin .field-label{margin-bottom:7px}:host.field-bottom-margin .field-message{min-height:14px;margin-top:5px}.field-label:empty{display:none}.field-message{font-size:75%;color:#0000008a}\n"], changeDetection: i0__namespace.ChangeDetectionStrategy.OnPush });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0__namespace, type: FsLabelFieldComponent, decorators: [{
                type: i0.Component,
                args: [{
                        selector: 'fs-label-field',
                        templateUrl: 'label-field.component.html',
                        styleUrls: ['label-field.component.scss'],
                        changeDetection: i0.ChangeDetectionStrategy.OnPush,
                    }]
            }], propDecorators: { setBottomMargin: [{
                    type: i0.Input,
                    args: ['bottomMargin']
                }], setTopMargin: [{
                    type: i0.Input,
                    args: ['topMargin']
                }], setLabelMargin: [{
                    type: i0.Input,
                    args: ['labelMargin']
                }], bottomMargin: [{
                    type: i0.HostBinding,
                    args: ['class.field-bottom-margin']
                }], topMargin: [{
                    type: i0.HostBinding,
                    args: ['class.field-top-margin']
                }], labelMargin: [{
                    type: i0.HostBinding,
                    args: ['class.field-label-margin']
                }] } });

    var FsLabelMessageComponent = /** @class */ (function () {
        function FsLabelMessageComponent() {
        }
        return FsLabelMessageComponent;
    }());
    FsLabelMessageComponent.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0__namespace, type: FsLabelMessageComponent, deps: [], target: i0__namespace.ɵɵFactoryTarget.Component });
    FsLabelMessageComponent.ɵcmp = i0__namespace.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.16", type: FsLabelMessageComponent, selector: "fs-label-message", ngImport: i0__namespace, template: "<ng-content></ng-content>", styles: [""], changeDetection: i0__namespace.ChangeDetectionStrategy.OnPush });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0__namespace, type: FsLabelMessageComponent, decorators: [{
                type: i0.Component,
                args: [{
                        selector: 'fs-label-message',
                        templateUrl: 'label-message.component.html',
                        styleUrls: ['label-message.component.scss'],
                        changeDetection: i0.ChangeDetectionStrategy.OnPush,
                    }]
            }], ctorParameters: function () { return []; } });

    var FsLabelModule = /** @class */ (function () {
        function FsLabelModule() {
        }
        return FsLabelModule;
    }());
    FsLabelModule.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0__namespace, type: FsLabelModule, deps: [], target: i0__namespace.ɵɵFactoryTarget.NgModule });
    FsLabelModule.ɵmod = i0__namespace.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0__namespace, type: FsLabelModule, declarations: [FsLabelComponent,
            FsLabelFieldComponent,
            FsLabelMessageComponent], imports: [common.CommonModule], exports: [FsLabelComponent,
            FsLabelFieldComponent,
            FsLabelMessageComponent] });
    FsLabelModule.ɵinj = i0__namespace.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0__namespace, type: FsLabelModule, imports: [[
                common.CommonModule,
            ]] });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0__namespace, type: FsLabelModule, decorators: [{
                type: i0.NgModule,
                args: [{
                        imports: [
                            common.CommonModule,
                        ],
                        exports: [
                            FsLabelComponent,
                            FsLabelFieldComponent,
                            FsLabelMessageComponent
                        ],
                        entryComponents: [],
                        declarations: [
                            FsLabelComponent,
                            FsLabelFieldComponent,
                            FsLabelMessageComponent
                        ]
                    }]
            }] });

    /*
     * Public API Surface of fs-menu
     */

    /**
     * Generated bundle index. Do not edit.
     */

    exports.FsLabelComponent = FsLabelComponent;
    exports.FsLabelFieldComponent = FsLabelFieldComponent;
    exports.FsLabelMessageComponent = FsLabelMessageComponent;
    exports.FsLabelModule = FsLabelModule;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
