// @firestitch/common/esm2015/app/fs-common.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FsArray } from './services/array.service';
import { FsMath } from './services/math.service';
import { FsUtil } from './services/util.service';
import { FsValidate } from './services/vaildate.service';
import { FsUtilGuidPipe, FsUtilStringifyPipe, FsUtilIsEmptyPipe, FsUtilIsNotEmptyPipe, } from './pipes/util.pipe';
import { FsFormatNumberPipe } from './pipes/number.pipe';
import { FsFormatCurrencyPipe } from './pipes/currency.pipe';
import { FsFormatTruncatePipe } from './pipes/truncate.pipe';
import { FsFormatBytesPipe } from './pipes/bytes.pipe';
import { FsStopPropagationDirective } from './directives/stop-propagation.directive';
import { FsPreventDefaultDirective } from './directives/prevent-default.directive';
import { FsAutofocusDirective } from './directives/autofocus.directive';
import { FsFormatOrdinalNumberPipe } from './pipes/ordinal-number.pipe';
import * as i0 from "@angular/core";
export class FsCommonModule {
    static forRoot() {
        return {
            ngModule: FsCommonModule,
            providers: [
                FsArray,
                FsMath,
                FsUtil,
                FsValidate
            ]
        };
    }
}
FsCommonModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsCommonModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
FsCommonModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsCommonModule, declarations: [FsStopPropagationDirective,
        FsPreventDefaultDirective,
        FsAutofocusDirective,
        FsUtilGuidPipe,
        FsUtilIsEmptyPipe,
        FsUtilIsNotEmptyPipe,
        FsUtilStringifyPipe,
        FsFormatNumberPipe,
        FsFormatCurrencyPipe,
        FsFormatTruncatePipe,
        FsFormatBytesPipe,
        FsFormatOrdinalNumberPipe], imports: [CommonModule], exports: [FsStopPropagationDirective,
        FsPreventDefaultDirective,
        FsAutofocusDirective,
        FsUtilGuidPipe,
        FsUtilStringifyPipe,
        FsFormatNumberPipe,
        FsFormatCurrencyPipe,
        FsFormatTruncatePipe,
        FsUtilIsEmptyPipe,
        FsUtilIsNotEmptyPipe,
        FsFormatBytesPipe,
        FsFormatOrdinalNumberPipe] });
FsCommonModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsCommonModule, imports: [[
            CommonModule,
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsCommonModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                    ],
                    exports: [
                        FsStopPropagationDirective,
                        FsPreventDefaultDirective,
                        FsAutofocusDirective,
                        FsUtilGuidPipe,
                        FsUtilStringifyPipe,
                        FsFormatNumberPipe,
                        FsFormatCurrencyPipe,
                        FsFormatTruncatePipe,
                        FsUtilIsEmptyPipe,
                        FsUtilIsNotEmptyPipe,
                        FsFormatBytesPipe,
                        FsFormatOrdinalNumberPipe,
                    ],
                    declarations: [
                        FsStopPropagationDirective,
                        FsPreventDefaultDirective,
                        FsAutofocusDirective,
                        FsUtilGuidPipe,
                        FsUtilIsEmptyPipe,
                        FsUtilIsNotEmptyPipe,
                        FsUtilStringifyPipe,
                        FsFormatNumberPipe,
                        FsFormatCurrencyPipe,
                        FsFormatTruncatePipe,
                        FsFormatBytesPipe,
                        FsFormatOrdinalNumberPipe,
                    ],
                }]
        }] });
