// @firestitch/menu/esm2015/app/directives/menu-item/fs-menu-item.directive.js
import { EventEmitter, ChangeDetectorRef, Directive, Input, Output, ContentChildren, TemplateRef, Optional, SkipSelf, ContentChild, } from '@angular/core';
import { Subject } from 'rxjs';
import { FsGroupMenuItemTemplateDirective } from '../group-menu-item-template/fs-group-menu-item-template.directive';
import * as i0 from "@angular/core";
export class FsMenuItemDirective {
    constructor(cd, templateRef, parent) {
        this.cd = cd;
        this.templateRef = templateRef;
        this.parent = parent;
        this.ngClass = [];
        this.cssClass = '';
        this.cssId = '';
        this.hidden = false;
        this.dismissAfterClick = true;
        this.target = null;
        this.queryParams = {};
        this.click$ = new EventEmitter();
        this.itemChange$ = new Subject();
        this._isGroup = false;
    }
    set itemsElements(value) {
        this.childrenItems = value.toArray()
            .filter((child) => child !== this);
        this._isGroup = !!this.childrenItems;
        this.checkChildrenVisibility();
        this.itemChange$.next();
    }
    get isGroup() {
        return this._isGroup;
    }
    get visible() {
        if (this.groupHidden !== void 0) {
            return !this.groupHidden;
        }
        else {
            return !this.hidden;
        }
    }
    get groupItemTemplateRef() {
        return this._groupItemTemplateRef;
    }
    ngOnChanges(changes) {
        if (!this.isGroup && this.parent && changes.hidden && !changes.hidden.firstChange) {
            this.parent.checkChildrenVisibility();
        }
        this.itemChange$.next();
    }
    checkChildrenVisibility() {
        if (this.childrenItems) {
            this.hidden = this.childrenItems.every((item) => item.hidden);
        }
    }
    ngOnDestroy() {
        this.itemChange$.complete();
    }
    click(event) {
        this.click$.next(event);
    }
}
/** @nocollapse */ FsMenuItemDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.13", ngImport: i0, type: FsMenuItemDirective, deps: [{ token: i0.ChangeDetectorRef }, { token: i0.TemplateRef, optional: true }, { token: FsMenuItemDirective, optional: true, skipSelf: true }], target: i0.ɵɵFactoryTarget.Directive });
/** @nocollapse */ FsMenuItemDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.13", type: FsMenuItemDirective, selector: "fs-menu-group,[fs-menu-item]", inputs: { ngClass: ["fsClass", "ngClass"], cssClass: ["class", "cssClass"], cssId: ["id", "cssId"], label: "label", hidden: "hidden", groupHidden: "groupHidden", dismissAfterClick: "dismissAfterClick", link: "link", target: "target", queryParams: "queryParams" }, outputs: { click$: "click" }, queries: [{ propertyName: "_groupItemTemplateRef", first: true, predicate: FsGroupMenuItemTemplateDirective, descendants: true, read: TemplateRef }, { propertyName: "itemsElements", predicate: FsMenuItemDirective }], usesOnChanges: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.13", ngImport: i0, type: FsMenuItemDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: 'fs-menu-group,[fs-menu-item]'
                }]
        }], ctorParameters: function () { return [{ type: i0.ChangeDetectorRef }, { type: i0.TemplateRef, decorators: [{
                    type: Optional
                }] }, { type: FsMenuItemDirective, decorators: [{
                    type: SkipSelf
                }, {
                    type: Optional
                }] }]; }, propDecorators: { ngClass: [{
                type: Input,
                args: ['fsClass']
            }], cssClass: [{
                type: Input,
                args: ['class']
            }], cssId: [{
                type: Input,
                args: ['id']
            }], label: [{
                type: Input
            }], hidden: [{
                type: Input
            }], groupHidden: [{
                type: Input
            }], dismissAfterClick: [{
                type: Input
            }], link: [{
                type: Input
            }], target: [{
                type: Input
            }], queryParams: [{
                type: Input
            }], click$: [{
                type: Output,
                args: ['click']
            }], itemsElements: [{
                type: ContentChildren,
                args: [FsMenuItemDirective]
            }], _groupItemTemplateRef: [{
                type: ContentChild,
                args: [FsGroupMenuItemTemplateDirective, { read: TemplateRef }]
            }] } });
