// @firestitch/drawer-unique-namespace/esm5/app/classes/drawer-data.js
import { Observable, Subject } from 'rxjs';
import { takeUntil } from 'rxjs/operators';
var DrawerData = /** @class */ (function () {
    function DrawerData(data) {
        if (data === void 0) { data = {}; }
        this._dataChange = new Subject();
        this._destroy = new Subject();
        this._data = data;
    }
    DrawerData.createWithProxy = function (data) {
        if (data === void 0) { data = {}; }
        var drawerData = new DrawerData(data);
        return new Proxy(drawerData, {
            get: function (target, property) {
                if (typeof target[property] === 'function'
                    || property === '_data'
                    || target[property] instanceof Observable) {
                    return target[property];
                }
                else {
                    return target._data[property];
                }
            },
            set: function (target, property, value) {
                if (property !== '_data') {
                    target._data[property] = value;
                }
                else {
                    target[property] = value;
                }
                return true;
            },
            has: function (target, property) {
                return property in target._data;
            },
            ownKeys: function (target) {
                return Object.keys(target._data);
            },
            enumerate: function (target) {
                return Object.keys(target._data);
            },
            getOwnPropertyDescriptor: function (target, property) {
                return Object.getOwnPropertyDescriptor(target._data, property);
            }
        });
    };
    Object.defineProperty(DrawerData.prototype, "dataChange$", {
        get: function () {
            return this._dataChange.pipe(takeUntil(this._destroy));
        },
        enumerable: true,
        configurable: true
    });
    DrawerData.prototype.getValue = function () {
        return this._data;
    };
    DrawerData.prototype.setValue = function (value) {
        this._data = value;
        this._dataChange.next(this._data);
    };
    DrawerData.prototype.destroy = function () {
        this._destroy.next();
        this._destroy.complete();
    };
    return DrawerData;
}());
export { DrawerData };
