// @firestitch/drawer-unique-namespace/esm5/app/classes/drawer-size-controller.js
import { __decorate, __metadata } from "tslib";
import { Injectable, NgZone, OnDestroy } from '@angular/core';
import { fromEvent, Subject } from 'rxjs';
import { debounceTime, takeUntil } from 'rxjs/operators';
import { DrawerRef } from '../classes/drawer-ref';
var MAIN_DRAWER_DEFAULT_WIDTH = 500;
var SIDE_DRAWER_DEFAULT_WIDTH = 200;
var SIDE_RESIZE_BAR_WIDTH = 25;
var MAIN_RESIZE_ACTION_BAR_WIDTH = 40;
var DrawerSizeController = /** @class */ (function () {
    function DrawerSizeController(_drawerRef, _ngZone) {
        this._drawerRef = _drawerRef;
        this._ngZone = _ngZone;
        this._sideOpened = false;
        this._borderPadding = 0;
        this._destroy$ = new Subject();
        this._initDefaultConfigs();
        this._updateScreenWidth();
        this._listenWindowResize();
        this._listenSideToggle();
    }
    Object.defineProperty(DrawerSizeController.prototype, "mainElRef", {
        get: function () {
            return this._mainElRef;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DrawerSizeController.prototype, "sideElRef", {
        get: function () {
            return this._sideElRef;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DrawerSizeController.prototype, "mainConfig", {
        get: function () {
            return this._mainConfig;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DrawerSizeController.prototype, "sideConfig", {
        get: function () {
            return this._sideConfig;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DrawerSizeController.prototype, "screenWidth", {
        get: function () {
            return this._screenWidth;
        },
        enumerable: true,
        configurable: true
    });
    DrawerSizeController.prototype.ngOnDestroy = function () {
        this._destroy$.next();
        this._destroy$.complete();
    };
    DrawerSizeController.prototype.registerElRef = function (el) {
        if (el.type === 'main') {
            this._registerMainRef(el);
        }
        else if (el.type === 'side') {
            this._registerSideRef(el);
        }
        else {
            throw Error('Unrecognized resize element type');
        }
    };
    DrawerSizeController.prototype.getInitialWidth = function (type) {
        if (type === 'main') {
            return this.mainConfig.initial;
        }
        else if (type === 'side') {
            return this.sideConfig.initial;
        }
        else {
            return void 0;
        }
    };
    DrawerSizeController.prototype.getMinWidth = function (type) {
        if (type === 'main') {
            return this.mainConfig.min;
        }
        else if (type === 'side') {
            return this.sideConfig.min;
        }
        else {
            return void 0;
        }
    };
    DrawerSizeController.prototype.getMaxWidth = function (type) {
        if (type === 'main') {
            return this.mainConfig.max;
        }
        else if (type === 'side') {
            return this.sideConfig.max;
        }
        else {
            return void 0;
        }
    };
    /**
     * Update width from outside with all calculations to be done
     * @param width
     */
    DrawerSizeController.prototype.updateMainWidth = function (width) {
        var sideWidth = (this.sideElRef && this.sideElRef.width) || 0;
        this.mainElRef.updateWidth(sideWidth + width);
    };
    /**
     * Update width from outside with all calculations to be done
     * @param width
     */
    DrawerSizeController.prototype.updateSideWidth = function (width) {
        if (this.sideElRef) {
            var currentWidth = this.mainElRef.width - this.sideElRef.width;
            this.mainElRef.updateWidth(currentWidth + width);
            this.sideElRef.updateWidth(width);
        }
    };
    /**
     * Push current drawer to be visible under new one opened
     * @param inFrontDrawer
     */
    DrawerSizeController.prototype.pushMainWidth = function (inFrontDrawer) {
        var inFrontDrawerTotalWidth = inFrontDrawer.resizeController.mainElRef.width + MAIN_RESIZE_ACTION_BAR_WIDTH;
        if (this.mainElRef.width <= inFrontDrawerTotalWidth) {
            this.updateMainWidth(inFrontDrawerTotalWidth);
        }
    };
    /**
     * Listen for browser resize and update restrictions
     */
    DrawerSizeController.prototype._listenWindowResize = function () {
        var _this = this;
        this._ngZone.runOutsideAngular(function () {
            fromEvent(window, 'resize')
                .pipe(debounceTime(50), takeUntil(_this._destroy$))
                .subscribe(function () {
                _this._updateScreenWidth();
                _this._updateMinMaxStyles();
            });
        });
    };
    /**
     * Copy initial configs or set default values
     */
    DrawerSizeController.prototype._initDefaultConfigs = function () {
        this._mainConfig =
            (this._drawerRef.drawerConfig.width && this._drawerRef.drawerConfig.width.main)
                || {};
        this._mainConfig.initial = this._mainConfig.initial || MAIN_DRAWER_DEFAULT_WIDTH;
        this._sideConfig =
            (this._drawerRef.drawerConfig.width && this._drawerRef.drawerConfig.width.side)
                || {};
        this._sideConfig.initial = this._sideConfig.initial || SIDE_DRAWER_DEFAULT_WIDTH;
    };
    DrawerSizeController.prototype._registerMainRef = function (el) {
        this._mainElRef = el;
    };
    DrawerSizeController.prototype._registerSideRef = function (el) {
        this._sideElRef = el;
    };
    /**
     * Update current window size
     */
    DrawerSizeController.prototype._updateScreenWidth = function () {
        this._screenWidth = (window.innerWidth - this._borderPadding);
    };
    /**
     * Update min&max css props for containers
     */
    DrawerSizeController.prototype._updateMinMaxStyles = function () {
        this.mainElRef.setMinMaxStyles();
        if (this.sideElRef) {
            this.sideElRef.setMinMaxStyles();
        }
    };
    DrawerSizeController.prototype._listenSideToggle = function () {
        var _this = this;
        this._drawerRef.sideToggle$
            .pipe(takeUntil(this._destroy$))
            .subscribe(function (opened) {
            if (_this._sideOpened === opened) {
                return;
            }
            _this._sideOpened = opened;
            if (opened) {
                var currentWidth = _this.mainElRef.width;
                var sideWidth = _this.getInitialWidth('side');
                _this._mainElRef.updateWidth(currentWidth + sideWidth + SIDE_RESIZE_BAR_WIDTH);
            }
            else {
                var actualSideWidth = _this.sideElRef.fsDrawerResizer.getBoundingClientRect().width;
                var mainWidth = _this.mainElRef.width - actualSideWidth - SIDE_RESIZE_BAR_WIDTH;
                _this._mainElRef.updateWidth(mainWidth);
            }
        });
    };
    DrawerSizeController.ctorParameters = function () { return [
        { type: DrawerRef },
        { type: NgZone }
    ]; };
    DrawerSizeController = __decorate([
        Injectable(),
        __metadata("design:paramtypes", [DrawerRef,
            NgZone])
    ], DrawerSizeController);
    return DrawerSizeController;
}());
export { DrawerSizeController };
