// @firestitch/drawer/esm2015/app/components/drawer-content/drawer-content.component.js
import { Component } from '@angular/core';
import * as i0 from "@angular/core";
export class FsDrawerContentComponent {
}
FsDrawerContentComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsDrawerContentComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
FsDrawerContentComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.16", type: FsDrawerContentComponent, selector: "[fsDrawerContent]", host: { classAttribute: "content-container" }, ngImport: i0, template: '<div class="content"><ng-content></ng-content></div>', isInline: true });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsDrawerContentComponent, decorators: [{
            type: Component,
            args: [{
                    selector: '[fsDrawerContent]',
                    template: '<div class="content"><ng-content></ng-content></div>',
                    host: {
                        'class': 'content-container'
                    }
                }]
        }] });
