// @firestitch/label/esm2015/app/fs-label.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FsLabelComponent } from './components/label/label.component';
import { FsLabelFieldComponent } from './components/label-field/label-field.component';
import { FsLabelMessageComponent } from './components/label-message/label-message.component';
import * as i0 from "@angular/core";
export class FsLabelModule {
}
FsLabelModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsLabelModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
FsLabelModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsLabelModule, declarations: [FsLabelComponent,
        FsLabelFieldComponent,
        FsLabelMessageComponent], imports: [CommonModule], exports: [FsLabelComponent,
        FsLabelFieldComponent,
        FsLabelMessageComponent] });
FsLabelModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsLabelModule, imports: [[
            CommonModule,
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsLabelModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                    ],
                    exports: [
                        FsLabelComponent,
                        FsLabelFieldComponent,
                        FsLabelMessageComponent
                    ],
                    entryComponents: [],
                    declarations: [
                        FsLabelComponent,
                        FsLabelFieldComponent,
                        FsLabelMessageComponent
                    ]
                }]
        }] });
