// @firestitch/drawer-unique-namespace/esm5/app/components/drawer-menu/drawer-menu.component.js
import { __decorate, __extends, __metadata } from "tslib";
import { ChangeDetectionStrategy, Component, ViewChild, } from '@angular/core';
import { BasePortalOutlet, CdkPortalOutlet } from '@angular/cdk/portal';
var FsDrawerMenuComponent = /** @class */ (function (_super) {
    __extends(FsDrawerMenuComponent, _super);
    function FsDrawerMenuComponent() {
        return _super.call(this) || this;
    }
    FsDrawerMenuComponent.prototype.setDrawerMenuRef = function (value) {
        this.ref = value;
        // this.config = this.externalMenuRef.config;
    };
    /**
     * Attach a ComponentPortal as content to this dialog container.
     * @param portal Portal to be attached as the dialog content.
     */
    FsDrawerMenuComponent.prototype.attachComponentPortal = function (portal) {
        if (this._portalOutlet.hasAttached()) {
            throw Error('Drawer component already attached');
        }
        return this._portalOutlet.attachComponentPortal(portal);
    };
    /**
     * Attach a TemplatePortal as content to this dialog container.
     * @param portal Portal to be attached as the dialog content.
     */
    FsDrawerMenuComponent.prototype.attachTemplatePortal = function (portal) {
        if (this._portalOutlet.hasAttached()) {
            throw Error('Drawer template already attached');
        }
        return this._portalOutlet.attachTemplatePortal(portal);
    };
    __decorate([
        ViewChild(CdkPortalOutlet, { static: true }),
        __metadata("design:type", CdkPortalOutlet)
    ], FsDrawerMenuComponent.prototype, "_portalOutlet", void 0);
    FsDrawerMenuComponent = __decorate([
        Component({
            selector: 'fs-drawer-menu',
            template: "<ng-template cdkPortalOutlet></ng-template>\n",
            host: {
                'class': 'mat-elevation-z2'
            },
            changeDetection: ChangeDetectionStrategy.OnPush,
            styles: [":host{padding:15px;background:#fff;border-radius:2px;overflow:auto}"]
        }),
        __metadata("design:paramtypes", [])
    ], FsDrawerMenuComponent);
    return FsDrawerMenuComponent;
}(BasePortalOutlet));
export { FsDrawerMenuComponent };
