// @firestitch/iframe/esm2015/app/directives/fs-iframe-scrollto/fs-iframe-scrollto.directive.js
import { Directive, ElementRef, HostListener, Renderer2 } from '@angular/core';
import { postMessage } from '../../functions/post-message';
import * as i0 from "@angular/core";
export class FsIFrameScrollToDirective {
    constructor(elementRef, renderer) {
        this.elementRef = elementRef;
        this.renderer = renderer;
    }
    ngOnInit() {
        this.renderer.addClass(this.elementRef.nativeElement, 'fs-iframe-scrollto');
    }
    onClick(event) {
        postMessage({ name: 'scrollto', value: this.elementRef.nativeElement.getBoundingClientRect().top });
    }
}
FsIFrameScrollToDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsIFrameScrollToDirective, deps: [{ token: i0.ElementRef }, { token: i0.Renderer2 }], target: i0.ɵɵFactoryTarget.Directive });
FsIFrameScrollToDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.16", type: FsIFrameScrollToDirective, selector: "[fsIFrameScrollTo]", host: { listeners: { "click": "onClick($event)" } }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsIFrameScrollToDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[fsIFrameScrollTo]'
                }]
        }], ctorParameters: function () { return [{ type: i0.ElementRef }, { type: i0.Renderer2 }]; }, propDecorators: { onClick: [{
                type: HostListener,
                args: ['click', ['$event']]
            }] } });
