// @firestitch/iframe/esm2015/app/fs-iframe.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FsIFrameScrollToDirective } from './directives/fs-iframe-scrollto/fs-iframe-scrollto.directive';
import { FsIFrameDirective } from './directives/fs-iframe/fs-iframe.directive';
import { FsIFrame } from './services/iframe.service';
import { FsIFrameComponent } from './../app/components/iframe/iframe.component';
import * as i0 from "@angular/core";
export class FsIFrameModule {
    static forRoot() {
        return {
            ngModule: FsIFrameModule,
            providers: [FsIFrame]
        };
    }
}
FsIFrameModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsIFrameModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
FsIFrameModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsIFrameModule, declarations: [FsIFrameScrollToDirective,
        FsIFrameDirective,
        FsIFrameComponent], imports: [CommonModule], exports: [FsIFrameScrollToDirective,
        FsIFrameDirective,
        FsIFrameComponent] });
FsIFrameModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsIFrameModule, imports: [[
            CommonModule,
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsIFrameModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                    ],
                    exports: [
                        FsIFrameScrollToDirective,
                        FsIFrameDirective,
                        FsIFrameComponent,
                    ],
                    declarations: [
                        FsIFrameScrollToDirective,
                        FsIFrameDirective,
                        FsIFrameComponent,
                    ],
                }]
        }] });
