// @firestitch/menu/esm2015/app/fs-menu.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { MatBottomSheetModule } from '@angular/material/bottom-sheet';
import { MatButtonModule } from '@angular/material/button';
import { MatIconModule } from '@angular/material/icon';
import { MatListModule } from '@angular/material/list';
import { MatMenuModule } from '@angular/material/menu';
import { RouterModule } from '@angular/router';
import { FsMenuComponent } from './components/fs-menu/fs-menu.component';
import { FsBottomSheetComponent } from './components/fs-menu/bottom-sheet/fs-bottom-sheet.component';
import { MenuItemsListComponent } from './components/fs-menu/menu-items-list/menu-items-list.component';
import { BottomItemsListComponent } from './components/fs-menu/bottom-sheet/bottom-items-list/bottom-items-list.component';
import { FsMenuItemDirective } from './directives/menu-item/fs-menu-item.directive';
import { FsMenuTriggerDirective } from './directives/menu-trigger/fs-menu-trigger.directive';
import { FsMenuTitleDirective } from './directives/menu-title/fs-menu-title.directive';
import { FsGroupMenuItemTemplateDirective } from './directives/group-menu-item-template/fs-group-menu-item-template.directive';
import * as i0 from "@angular/core";
export class FsMenuModule {
    static forRoot() {
        return {
            ngModule: FsMenuModule
        };
    }
}
/** @nocollapse */ FsMenuModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.13", ngImport: i0, type: FsMenuModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
/** @nocollapse */ FsMenuModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.13", ngImport: i0, type: FsMenuModule, declarations: [FsMenuComponent,
        FsBottomSheetComponent,
        MenuItemsListComponent,
        BottomItemsListComponent,
        FsMenuItemDirective,
        FsMenuTriggerDirective,
        FsMenuTitleDirective,
        FsGroupMenuItemTemplateDirective], imports: [CommonModule,
        RouterModule,
        MatButtonModule,
        MatIconModule,
        MatMenuModule,
        MatBottomSheetModule,
        MatListModule], exports: [FsMenuComponent,
        FsMenuItemDirective,
        FsMenuTriggerDirective,
        FsMenuTitleDirective,
        FsGroupMenuItemTemplateDirective] });
/** @nocollapse */ FsMenuModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.13", ngImport: i0, type: FsMenuModule, imports: [[
            CommonModule,
            RouterModule,
            MatButtonModule,
            MatIconModule,
            MatMenuModule,
            MatBottomSheetModule,
            MatListModule,
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.13", ngImport: i0, type: FsMenuModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        RouterModule,
                        MatButtonModule,
                        MatIconModule,
                        MatMenuModule,
                        MatBottomSheetModule,
                        MatListModule,
                    ],
                    exports: [
                        FsMenuComponent,
                        FsMenuItemDirective,
                        FsMenuTriggerDirective,
                        FsMenuTitleDirective,
                        FsGroupMenuItemTemplateDirective,
                    ],
                    declarations: [
                        FsMenuComponent,
                        FsBottomSheetComponent,
                        MenuItemsListComponent,
                        BottomItemsListComponent,
                        FsMenuItemDirective,
                        FsMenuTriggerDirective,
                        FsMenuTitleDirective,
                        FsGroupMenuItemTemplateDirective,
                    ]
                }]
        }] });
