// @firestitch/common/esm2015/app/pipes/currency.pipe.js
import { Pipe, LOCALE_ID, Inject } from '@angular/core';
import { currency } from '../../libs/format/currency/currency';
import * as i0 from "@angular/core";
export class FsFormatCurrencyPipe {
    constructor(locale) {
        this.locale = locale;
    }
    transform(amount, precision, currencyCode) {
        return currency(amount, precision, currencyCode);
    }
}
FsFormatCurrencyPipe.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsFormatCurrencyPipe, deps: [{ token: LOCALE_ID }], target: i0.ɵɵFactoryTarget.Pipe });
FsFormatCurrencyPipe.ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsFormatCurrencyPipe, name: "fsFormatCurrency" });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsFormatCurrencyPipe, decorators: [{
            type: Pipe,
            args: [{
                    name: 'fsFormatCurrency'
                }]
        }], ctorParameters: function () { return [{ type: undefined, decorators: [{
                    type: Inject,
                    args: [LOCALE_ID]
                }] }]; } });
