// @firestitch/common/esm2015/app/pipes/util.pipe.js
import { Pipe } from '@angular/core';
import { stringify } from '../../libs/util/stringify/stringify';
import { guid } from '../../libs/util/guid/guid';
import { isEmpty } from '../../libs/util/isEmpty/isEmpty';
import * as i0 from "@angular/core";
export class FsUtilStringifyPipe {
    constructor() { }
    transform(value) {
        return stringify(value);
    }
}
FsUtilStringifyPipe.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsUtilStringifyPipe, deps: [], target: i0.ɵɵFactoryTarget.Pipe });
FsUtilStringifyPipe.ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsUtilStringifyPipe, name: "fsUtilStringify" });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsUtilStringifyPipe, decorators: [{
            type: Pipe,
            args: [{
                    name: 'fsUtilStringify'
                }]
        }], ctorParameters: function () { return []; } });
export class FsUtilGuidPipe {
    constructor() { }
    transform() {
        return guid();
    }
}
FsUtilGuidPipe.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsUtilGuidPipe, deps: [], target: i0.ɵɵFactoryTarget.Pipe });
FsUtilGuidPipe.ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsUtilGuidPipe, name: "fsUtilGuid" });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsUtilGuidPipe, decorators: [{
            type: Pipe,
            args: [{
                    name: 'fsUtilGuid'
                }]
        }], ctorParameters: function () { return []; } });
export class FsUtilIsEmptyPipe {
    transform(value) {
        return isEmpty(value);
    }
}
FsUtilIsEmptyPipe.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsUtilIsEmptyPipe, deps: [], target: i0.ɵɵFactoryTarget.Pipe });
FsUtilIsEmptyPipe.ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsUtilIsEmptyPipe, name: "fsUtilIsEmpty" });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsUtilIsEmptyPipe, decorators: [{
            type: Pipe,
            args: [{
                    name: 'fsUtilIsEmpty'
                }]
        }] });
export class FsUtilIsNotEmptyPipe {
    transform(value) {
        return !isEmpty(value);
    }
}
FsUtilIsNotEmptyPipe.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsUtilIsNotEmptyPipe, deps: [], target: i0.ɵɵFactoryTarget.Pipe });
FsUtilIsNotEmptyPipe.ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsUtilIsNotEmptyPipe, name: "fsUtilIsNotEmpty" });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsUtilIsNotEmptyPipe, decorators: [{
            type: Pipe,
            args: [{
                    name: 'fsUtilIsNotEmpty'
                }]
        }] });
