// @firestitch/menu/esm2015/app/directives/menu-trigger/fs-menu-trigger.directive.js
import { Directive, Input, Optional, Inject, Self, ElementRef, ViewContainerRef, } from '@angular/core';
import { MatMenuTrigger, MatMenu, MatMenuItem, MAT_MENU_SCROLL_STRATEGY } from '@angular/material/menu';
import { Directionality } from '@angular/cdk/bidi';
import { Overlay } from '@angular/cdk/overlay';
import { FocusMonitor } from '@angular/cdk/a11y';
import { FsMenuComponent } from '../../components/fs-menu/fs-menu.component';
import * as i0 from "@angular/core";
import * as i1 from "@angular/cdk/overlay";
import * as i2 from "@angular/material/menu";
import * as i3 from "@angular/cdk/bidi";
import * as i4 from "@angular/cdk/a11y";
export class FsMenuTriggerDirective extends MatMenuTrigger {
    constructor(_overlay, _element, _viewContainerRef, scrollStrategy, _parentMenu, _menuItemInstance, _dir, _focusMonitor) {
        super(_overlay, _element, _viewContainerRef, scrollStrategy, _parentMenu, _menuItemInstance, _dir, _focusMonitor);
        this.fsMenu = null;
    }
    ngOnInit() {
        this.menu = this.fsMenu.fsMenuRef;
        this.fsMenu.externalMatMenuTrigger = this;
    }
    _handleKeydown(event) {
        this._triggerClick();
    }
    /** Handles click events on the trigger. */
    _handleClick(event) {
        this._triggerClick();
    }
    _triggerClick() {
        if (this.fsMenu.menuOpened) {
            this.fsMenu.closeMenu();
        }
        else {
            this.fsMenu.openMenu();
        }
    }
}
/** @nocollapse */ FsMenuTriggerDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.13", ngImport: i0, type: FsMenuTriggerDirective, deps: [{ token: i1.Overlay }, { token: i0.ElementRef }, { token: i0.ViewContainerRef }, { token: MAT_MENU_SCROLL_STRATEGY }, { token: i2.MatMenu, optional: true }, { token: i2.MatMenuItem, optional: true, self: true }, { token: i3.Directionality, optional: true }, { token: i4.FocusMonitor }], target: i0.ɵɵFactoryTarget.Directive });
/** @nocollapse */ FsMenuTriggerDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.13", type: FsMenuTriggerDirective, selector: "[fsMenuTriggerFor]", inputs: { fsMenu: ["fsMenuTriggerFor", "fsMenu"] }, usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.13", ngImport: i0, type: FsMenuTriggerDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[fsMenuTriggerFor]',
                }]
        }], ctorParameters: function () { return [{ type: i1.Overlay }, { type: i0.ElementRef }, { type: i0.ViewContainerRef }, { type: undefined, decorators: [{
                    type: Inject,
                    args: [MAT_MENU_SCROLL_STRATEGY]
                }] }, { type: i2.MatMenu, decorators: [{
                    type: Optional
                }] }, { type: i2.MatMenuItem, decorators: [{
                    type: Optional
                }, {
                    type: Self
                }] }, { type: i3.Directionality, decorators: [{
                    type: Optional
                }] }, { type: i4.FocusMonitor }]; }, propDecorators: { fsMenu: [{
                type: Input,
                args: ['fsMenuTriggerFor']
            }] } });
