// @firestitch/message/esm2015/app/fs-message.module.js
import { NgModule, CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';
import { CommonModule } from '@angular/common';
import { MatButtonModule } from '@angular/material/button';
import { MatDialogModule } from '@angular/material/dialog';
import { MatIconModule } from '@angular/material/icon';
import { merge } from 'lodash-es';
import { FsMessageDialogComponent } from './components/message-dialog/message-dialog.component';
import { FsMessagesComponent } from './components/messages/messages.component';
import { FsMessageComponent } from './components/message/message.component';
import { FsMessage } from './message.service';
import { FS_MESSAGE_CONFIG, FS_MESSAGE_DEFAULT_CONFIG } from './injectors/message-config';
import { MessageMode } from './enums';
import * as i0 from "@angular/core";
export class FsMessageModule {
    static forRoot(config = {}) {
        return {
            ngModule: FsMessageModule,
            providers: [
                FsMessage,
                { provide: FS_MESSAGE_DEFAULT_CONFIG, useValue: config },
                {
                    provide: FS_MESSAGE_CONFIG,
                    useFactory: FsMessageConfigFactory,
                    deps: [FS_MESSAGE_DEFAULT_CONFIG]
                },
            ]
        };
    }
}
FsMessageModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsMessageModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
FsMessageModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsMessageModule, declarations: [FsMessagesComponent,
        FsMessageComponent,
        FsMessageDialogComponent], imports: [CommonModule,
        MatDialogModule,
        MatIconModule,
        MatButtonModule], exports: [FsMessagesComponent,
        FsMessageComponent] });
FsMessageModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsMessageModule, imports: [[
            CommonModule,
            MatDialogModule,
            MatIconModule,
            MatButtonModule
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsMessageModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        MatDialogModule,
                        MatIconModule,
                        MatButtonModule
                    ],
                    exports: [
                        FsMessagesComponent,
                        FsMessageComponent
                    ],
                    entryComponents: [
                        FsMessageDialogComponent
                    ],
                    declarations: [
                        FsMessagesComponent,
                        FsMessageComponent,
                        FsMessageDialogComponent
                    ],
                    schemas: [CUSTOM_ELEMENTS_SCHEMA]
                }]
        }] });
export function FsMessageConfigFactory(config) {
    return merge({
        toastTimeout: 5,
        bannerTimeout: 5,
        dialogWidth: '500px',
        successMode: MessageMode.Toast,
        errorMode: MessageMode.Dialog,
        warningMode: MessageMode.Toast,
        infoMode: MessageMode.Toast
    }, config);
}
