// @firestitch/iframe/esm2015/app/directives/fs-iframe/fs-iframe.directive.js
import { Directive, Renderer2, ElementRef, HostListener } from '@angular/core';
import { parseEvent } from '../../functions/parse-event';
import * as i0 from "@angular/core";
export class FsIFrameDirective {
    constructor(elementRef, renderer) {
        this.elementRef = elementRef;
        this.renderer = renderer;
        this.offset = 0;
    }
    ngOnInit() {
        this.renderer.addClass(this.elementRef.nativeElement, 'fs-iframe');
    }
    onMessage(event) {
        const data = parseEvent(event);
        if (data.name === 'height') {
            const height = data.value;
            this.renderer.setStyle(this.elementRef.nativeElement, 'height', height + 'px');
        }
        else if (data.name === 'scrollto') {
            window.document.documentElement.scrollTop = data.value + this.offset;
        }
        else if (data.name === 'offset') {
            this.offset = data.value;
        }
    }
}
FsIFrameDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsIFrameDirective, deps: [{ token: i0.ElementRef }, { token: i0.Renderer2 }], target: i0.ɵɵFactoryTarget.Directive });
FsIFrameDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.16", type: FsIFrameDirective, selector: "[fsIFrame]", host: { listeners: { "window:message": "onMessage($event)" } }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsIFrameDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[fsIFrame]'
                }]
        }], ctorParameters: function () { return [{ type: i0.ElementRef }, { type: i0.Renderer2 }]; }, propDecorators: { onMessage: [{
                type: HostListener,
                args: ['window:message', ['$event']]
            }] } });
