// @firestitch/iframe/esm2015/app/services/iframe.service.js
import { Injectable } from '@angular/core';
import { interval } from 'rxjs';
import { postMessage } from '../functions/post-message';
import * as i0 from "@angular/core";
export class FsIFrame {
    observeBody() {
        return interval(100)
            .subscribe(() => {
            const height = window.document.body.offsetHeight;
            if (this.height !== height) {
                this.updateHeight(height);
                this.height = height;
            }
        });
    }
    updateCurrentHeight() {
        this.updateHeight(window.document.body.offsetHeight);
    }
    updateHeight(value) {
        this.update('height', value);
    }
    update(name, value) {
        postMessage({ name: name, value: value });
    }
}
FsIFrame.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsIFrame, deps: [], target: i0.ɵɵFactoryTarget.Injectable });
FsIFrame.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsIFrame });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsIFrame, decorators: [{
            type: Injectable
        }] });
