// @firestitch/message/esm2015/app/components/message/message.component.js
import { ChangeDetectionStrategy, Component, Input } from '@angular/core';
import { MessageType } from '../../enums';
import { FsMessage } from '../../message.service';
import * as i0 from "@angular/core";
import * as i1 from "../../message.service";
import * as i2 from "@angular/material/icon";
import * as i3 from "@angular/common";
export class FsMessageComponent {
    constructor(_fsMessage) {
        this._fsMessage = _fsMessage;
        this.type = MessageType.Info;
        this.icon = 'info';
    }
    set fsType(value) {
        this.type = value;
    }
    set fsMessage(value) {
        this.message = value;
    }
    ngOnInit() {
        this.icon = this._fsMessage.getIconName(this.type);
    }
}
FsMessageComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsMessageComponent, deps: [{ token: i1.FsMessage }], target: i0.ɵɵFactoryTarget.Component });
FsMessageComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.16", type: FsMessageComponent, selector: "fs-message", inputs: { type: "type", message: "message", fsType: "fsType", fsMessage: "fsMessage" }, ngImport: i0, template: "<div class=\"fs-message-banner fs-message-{{ type }}\">\n  <mat-icon *ngIf=\"icon\" class=\"mat-icon material-icons\" role=\"img\">{{ icon }}</mat-icon>\n  <div class=\"message\">\n    {{ message }}\n    <ng-content></ng-content>\n  </div>\n</div>\n", components: [{ type: i2.MatIcon, selector: "mat-icon", inputs: ["color", "inline", "svgIcon", "fontSet", "fontIcon"], exportAs: ["matIcon"] }], directives: [{ type: i3.NgIf, selector: "[ngIf]", inputs: ["ngIf", "ngIfThen", "ngIfElse"] }], changeDetection: i0.ChangeDetectionStrategy.OnPush });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsMessageComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'fs-message',
                    templateUrl: './message.component.html',
                    changeDetection: ChangeDetectionStrategy.OnPush,
                }]
        }], ctorParameters: function () { return [{ type: i1.FsMessage }]; }, propDecorators: { type: [{
                type: Input
            }], message: [{
                type: Input
            }], fsType: [{
                type: Input,
                args: ['fsType']
            }], fsMessage: [{
                type: Input,
                args: ['fsMessage']
            }] } });
