// @firestitch/label/esm2015/app/components/label-message/label-message.component.js
import { ChangeDetectionStrategy, Component } from '@angular/core';
import * as i0 from "@angular/core";
export class FsLabelMessageComponent {
    constructor() {
    }
}
FsLabelMessageComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsLabelMessageComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
FsLabelMessageComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.16", type: FsLabelMessageComponent, selector: "fs-label-message", ngImport: i0, template: "<ng-content></ng-content>", styles: [""], changeDetection: i0.ChangeDetectionStrategy.OnPush });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsLabelMessageComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'fs-label-message',
                    templateUrl: 'label-message.component.html',
                    styleUrls: ['label-message.component.scss'],
                    changeDetection: ChangeDetectionStrategy.OnPush,
                }]
        }], ctorParameters: function () { return []; } });
