// @firestitch/drawer-unique-namespace/esm5/app/components/drawer-actions/drawer-action-item/drawer-action-item.component.js
import { __decorate, __metadata } from "tslib";
import { ChangeDetectionStrategy, ChangeDetectorRef, Component, EventEmitter, Input, OnChanges, OnInit, Output, SimpleChanges } from '@angular/core';
import { Subject } from 'rxjs';
import { debounceTime, filter, takeUntil } from 'rxjs/operators';
import { DrawerRef } from '../../../classes/drawer-ref';
import { FsDrawerAction } from '../../../helpers/action-type.enum';
import { Action } from '../../../models/action.model';
var FsDrawerActionItemComponent = /** @class */ (function () {
    function FsDrawerActionItemComponent(drawer, _cdRef) {
        this.drawer = drawer;
        this._cdRef = _cdRef;
        this.actionClicked = new EventEmitter();
        this.isActive = false;
        this.actionTypes = FsDrawerAction;
        this._destroy$ = new Subject();
        this._listenActionChanges();
        this._listenDataChanges();
    }
    FsDrawerActionItemComponent.prototype.ngOnInit = function () {
        this._updateVisibilityAndLinks();
    };
    FsDrawerActionItemComponent.prototype.ngOnChanges = function (changes) {
        if (changes.activeAction) {
            this.isActive = this.action.name !== '' && this.action.name === this.activeAction;
        }
    };
    FsDrawerActionItemComponent.prototype.actionClick = function (event) {
        this.actionClicked.emit(event);
    };
    FsDrawerActionItemComponent.prototype.menuActionClick = function (action, event) {
        if (action.click) {
            action.click.call(null, {
                data: this.drawer.drawerData,
                event: event,
                drawerRef: this.drawer,
                action: action
            });
        }
    };
    FsDrawerActionItemComponent.prototype._listenActionChanges = function () {
        var _this = this;
        this.drawer.actionUpdated$
            .pipe(filter(function (name) { return name === _this.action.name; }), takeUntil(this._destroy$))
            .subscribe(function () {
            _this._cdRef.detectChanges();
        });
    };
    FsDrawerActionItemComponent.prototype._listenDataChanges = function () {
        var _this = this;
        this.drawer.dataChanged$
            .pipe(takeUntil(this._destroy$), debounceTime(50))
            .subscribe(function () {
            _this._updateVisibilityAndLinks();
        });
    };
    FsDrawerActionItemComponent.prototype._updateVisibilityAndLinks = function () {
        this._updateVisibility();
        this._updateRouterLinks();
        this._cdRef.detectChanges();
    };
    FsDrawerActionItemComponent.prototype._updateVisibility = function () {
        this.action.checkVisibility(this.drawer.drawerData);
    };
    FsDrawerActionItemComponent.prototype._updateRouterLinks = function () {
        if (this.action.type === this.actionTypes.Menu) {
            this.action.updateRouterLink({
                data: this.drawer.drawerData,
                drawerRef: this.drawer
            });
        }
    };
    FsDrawerActionItemComponent.ctorParameters = function () { return [
        { type: DrawerRef },
        { type: ChangeDetectorRef }
    ]; };
    __decorate([
        Input(),
        __metadata("design:type", Action)
    ], FsDrawerActionItemComponent.prototype, "action", void 0);
    __decorate([
        Input(),
        __metadata("design:type", String)
    ], FsDrawerActionItemComponent.prototype, "activeAction", void 0);
    __decorate([
        Output(),
        __metadata("design:type", Object)
    ], FsDrawerActionItemComponent.prototype, "actionClicked", void 0);
    FsDrawerActionItemComponent = __decorate([
        Component({
            selector: 'fs-drawer-action-item',
            template: "<ng-container [ngSwitch]=\"action.type\">\n  <ng-container *ngSwitchCase=\"actionTypes.Button\">\n    <button mat-icon-button\n            type=\"button\"\n            *ngIf=\"action.visible\"\n            [matTooltip]=\"action.tooltip\"\n            [class.active]=\"isActive\"\n            (click)=\"actionClick($event)\">\n      <i class=\"material-icons\">{{action.icon}}</i>\n    </button>\n  </ng-container>\n\n  <ng-container *ngSwitchCase=\"actionTypes.Menu\">\n    <ng-container *ngIf=\"action.visible\">\n      <button mat-icon-button\n              type=\"button\"\n              class=\"sub-action\"\n              [fsMenuTriggerFor]=\"menu\"\n              [matTooltip]=\"action.tooltip\"\n              [class.active]=\"isActive\"\n              (click)=\"actionClick($event)\">\n        <i class=\"material-icons\">{{action.icon}}</i>\n      </button>\n\n      <fs-menu #menu>\n        <ng-container *ngFor=\"let menuAction of action.menuActions\">\n          <ng-container *ngIf=\"menuAction.isGroup else simpleMenuItem\">\n            <fs-menu-group [label]=\"menuAction.label\" [hidden]=\"!menuAction.visible\">\n              <ng-container *ngFor=\"let groupAction of menuAction.actions\">\n                <ng-template fs-menu-item\n                             [hidden]=\"!groupAction.visible\"\n                             [link]=\"groupAction.routerLink?.link\"\n                             [queryParams]=\"groupAction.routerLink?.queryParams\"\n                             (click)=\"menuActionClick(groupAction, $event);\">\n                  <mat-icon *ngIf=\"menuAction.icon\">{{groupAction.icon}}</mat-icon>\n                  {{groupAction.label}}\n                </ng-template>\n              </ng-container>\n            </fs-menu-group>\n          </ng-container>\n          <ng-template #simpleMenuItem>\n            <ng-template fs-menu-item\n                         [hidden]=\"!menuAction.visible\"\n                         [link]=\"menuAction.routerLink?.link\"\n                         [queryParams]=\"menuAction.routerLink?.queryParams\"\n                         (click)=\"menuActionClick(menuAction, $event);\">\n              <i class=\"material-icons\" *ngIf=\"menuAction.icon\">{{menuAction.icon}}</i>\n              <span>{{menuAction.label}}</span>\n            </ng-template>\n          </ng-template>\n        </ng-container>\n      </fs-menu>\n    </ng-container>\n  </ng-container>\n\n  <ng-container *ngSwitchCase=\"actionTypes.Component\">\n    <button mat-icon-button\n            type=\"button\"\n            *ngIf=\"action.visible\"\n            [matTooltip]=\"action.tooltip\"\n            [class.active]=\"isActive\"\n            (click)=\"actionClick($event)\">\n      <i class=\"material-icons\">{{action.icon}}</i>\n    </button>\n  </ng-container>\n</ng-container>\n",
            changeDetection: ChangeDetectionStrategy.OnPush
        }),
        __metadata("design:paramtypes", [DrawerRef,
            ChangeDetectorRef])
    ], FsDrawerActionItemComponent);
    return FsDrawerActionItemComponent;
}());
export { FsDrawerActionItemComponent };
