// @firestitch/label/esm2015/app/components/label/label.component.js
import { ChangeDetectionStrategy, Component } from '@angular/core';
import * as i0 from "@angular/core";
export class FsLabelComponent {
    constructor() {
    }
}
FsLabelComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsLabelComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
FsLabelComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.16", type: FsLabelComponent, selector: "fs-label", ngImport: i0, template: "<ng-content></ng-content>", styles: [":host{transform-origin:0 0;transition:none;display:flex;color:#0000008a;font-size:75%;transform:translateY(-.15em) perspective(100px);min-height:12px}\n"], changeDetection: i0.ChangeDetectionStrategy.OnPush });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsLabelComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'fs-label',
                    templateUrl: 'label.component.html',
                    styleUrls: ['label.component.scss'],
                    changeDetection: ChangeDetectionStrategy.OnPush,
                }]
        }], ctorParameters: function () { return []; } });
