// @firestitch/drawer/esm2015/app/services/drawer-store.service.js
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
export class DrawerStoreService {
    constructor() {
        this._openedDrawers = [];
    }
    get drawerRefs() {
        return [...this._openedDrawers];
    }
    get numberOfOpenedDrawers() {
        return this._openedDrawers.length;
    }
    getLevelForRef(value) {
        return this._openedDrawers.indexOf(value) + 1;
    }
    addRef(value) {
        if (this._openedDrawers.indexOf(value) === -1) {
            this._openedDrawers.push(value);
            this._pushDrawersCascade();
        }
    }
    deleteRef(value) {
        this._openedDrawers = this._openedDrawers
            .filter((ref) => ref !== value);
    }
    /**
     * In case, when we want to open more than 1 drawer
     * our previously opened drawers should be visible
     *
     *      d1   d2   d3
     *     ---- ---- ---
     *    | x  | x1 | x2
     *    | y  | y1 | y2
     *    | z  | z1 | z2
     *     ---- ---- ---
     *
     * Where d1, d2 - previously opened drawers
     * d1 and d2 must be pushed left to be visible under just opened d3
     */
    _pushDrawersCascade() {
        if (this.numberOfOpenedDrawers > 1) {
            // SetTimeout should be here because we must wait render newly opened drawer
            // to be able to get his width
            setTimeout(() => {
                const refsArr = Array.from(this._openedDrawers.values());
                for (let i = refsArr.length - 1; i > 0; i--) {
                    const prevRef = refsArr[i - 1];
                    const currRef = refsArr[i];
                    prevRef.resizeController.pushMainWidth(currRef);
                }
            });
        }
    }
}
DrawerStoreService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: DrawerStoreService, deps: [], target: i0.ɵɵFactoryTarget.Injectable });
DrawerStoreService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: DrawerStoreService, providedIn: 'root' });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: DrawerStoreService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root',
                }]
        }] });
