// @firestitch/label/esm2015/app/components/label-field/label-field.component.js
import { ChangeDetectionStrategy, Component, HostBinding, Input } from '@angular/core';
import * as i0 from "@angular/core";
export class FsLabelFieldComponent {
    constructor() {
        this.bottomMargin = true;
        this.topMargin = true;
        this.labelMargin = true;
    }
    set setBottomMargin(value) {
        this.bottomMargin = value;
    }
    set setTopMargin(value) {
        this.topMargin = value;
    }
    set setLabelMargin(value) {
        this.labelMargin = value;
    }
}
FsLabelFieldComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsLabelFieldComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
FsLabelFieldComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.16", type: FsLabelFieldComponent, selector: "fs-label-field", inputs: { setBottomMargin: ["bottomMargin", "setBottomMargin"], setTopMargin: ["topMargin", "setTopMargin"], setLabelMargin: ["labelMargin", "setLabelMargin"] }, host: { properties: { "class.field-bottom-margin": "this.bottomMargin", "class.field-top-margin": "this.topMargin", "class.field-label-margin": "this.labelMargin" } }, ngImport: i0, template: "<div class=\"field-wrap\">\n  <div class=\"field-label\"><ng-content select=\"fs-label\"></ng-content></div>\n\n  <div class=\"field-content\">\n    <ng-content></ng-content>\n  </div>\n</div>\n\n<div class=\"field-message\">\n  <ng-content select=\"fs-label-message\"></ng-content>\n</div>\n", styles: [":host{display:block;position:relative}:host.field-label-margin .field-label{margin-bottom:7px}:host.field-bottom-margin .field-message{min-height:14px;margin-top:5px}.field-label:empty{display:none}.field-message{font-size:75%;color:#0000008a}\n"], changeDetection: i0.ChangeDetectionStrategy.OnPush });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsLabelFieldComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'fs-label-field',
                    templateUrl: 'label-field.component.html',
                    styleUrls: ['label-field.component.scss'],
                    changeDetection: ChangeDetectionStrategy.OnPush,
                }]
        }], propDecorators: { setBottomMargin: [{
                type: Input,
                args: ['bottomMargin']
            }], setTopMargin: [{
                type: Input,
                args: ['topMargin']
            }], setLabelMargin: [{
                type: Input,
                args: ['labelMargin']
            }], bottomMargin: [{
                type: HostBinding,
                args: ['class.field-bottom-margin']
            }], topMargin: [{
                type: HostBinding,
                args: ['class.field-top-margin']
            }], labelMargin: [{
                type: HostBinding,
                args: ['class.field-label-margin']
            }] } });
