// @firestitch/iframe/esm2015/app/components/iframe/iframe.component.js
import { Component, ElementRef, Input, ViewChild } from '@angular/core';
import { debounceTime, takeUntil } from 'rxjs/operators';
import { Subject, fromEvent } from 'rxjs';
import * as i0 from "@angular/core";
export class FsIFrameComponent {
    constructor() {
        this._destroy$ = new Subject();
    }
    onload() {
        this.updateHeight();
    }
    get frameEl() {
        var _a;
        return (_a = this.frame) === null || _a === void 0 ? void 0 : _a.nativeElement;
    }
    ngAfterViewInit() {
        fromEvent(window, 'resize')
            .pipe(debounceTime(50), takeUntil(this._destroy$))
            .subscribe(() => {
            this.updateHeight();
        });
        if (this.html) {
            this._updateBodyFrames();
        }
    }
    _updateBodyFrames() {
        const win = this.frameEl.contentWindow;
        const doc = win.document;
        const data = `
    <style>
      body {
        font-family: Roboto;
        font-size: 15px;
        margin: 0 !important;
        overflow-y: hidden !important;
        width: auto !important;
      }

      a {
        color: #1155CC;
      }

      * {
        box-sizing: border-box !important;
      }

      ${this.styles}
    </style>
    ${this.html}`;
        doc.open();
        doc.write(data);
        doc.close();
    }
    ngOnDestroy() {
        this._destroy$.next();
        this._destroy$.complete();
    }
    updateHeight() {
        if (this.frameEl) {
            this.frameEl.removeAttribute('height');
            this.frameEl.setAttribute('height', this.frameEl.contentDocument.body.scrollHeight);
        }
    }
}
FsIFrameComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsIFrameComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
FsIFrameComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.16", type: FsIFrameComponent, selector: "fs-iframe", inputs: { html: "html", styles: "styles" }, viewQueries: [{ propertyName: "frame", first: true, predicate: ["frame"], descendants: true }], ngImport: i0, template: "<iframe #frame (load)=\"onload()\"></iframe>\n", styles: ["iframe{width:100%;border:none}\n"] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsIFrameComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'fs-iframe',
                    templateUrl: './iframe.component.html',
                    styleUrls: ['./iframe.component.scss'],
                }]
        }], propDecorators: { frame: [{
                type: ViewChild,
                args: ['frame']
            }], html: [{
                type: Input
            }], styles: [{
                type: Input
            }] } });
