// @firestitch/example/esm2015/app/fs-example.module.js
import { NgModule, Inject } from '@angular/core';
import { CommonModule } from '@angular/common';
import { HttpClientModule } from '@angular/common/http';
import { MatButtonModule } from '@angular/material/button';
import { MatCardModule } from '@angular/material/card';
import { MatIconModule } from '@angular/material/icon';
import { MatTabsModule } from '@angular/material/tabs';
import { MatToolbarModule } from '@angular/material/toolbar';
import { MatTooltipModule } from '@angular/material/tooltip';
import { FsIFrameModule, FsIFrame } from '@firestitch/iframe';
import { FsDrawerModule } from '@firestitch/drawer-unique-namespace';
import { HIGHLIGHT_OPTIONS, HighlightModule } from 'ngx-highlightjs';
import { FsExampleComponent } from './components/fs-example/fs-example.component';
import { FsExampleHighlightComponent } from './components/fs-example-highlight/fs-example-highlight.component';
import { FsExamplesComponent } from './components/fs-examples/fs-examples.component';
import { FsExampleService } from './services/fs-example.service';
import { CopierService } from './services/copier';
import { highlightOptionsFactory } from './factories/fact';
import * as i0 from "@angular/core";
import * as i1 from "@firestitch/iframe";
export class FsExampleModule {
    constructor(config, fsIFrame) {
        this.config = config;
        this.fsIFrame = fsIFrame;
        if (!config) {
            config = {};
        }
        if (config.iframeObserveBody !== false) {
            fsIFrame.observeBody();
        }
    }
    static forRoot(config) {
        return {
            ngModule: FsExampleModule,
            providers: [
                {
                    provide: HIGHLIGHT_OPTIONS,
                    useFactory: highlightOptionsFactory,
                },
                { provide: 'FS_EXAMPLE_CONFIG', useValue: config },
                FsExampleService,
                CopierService,
            ]
        };
    }
}
FsExampleModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsExampleModule, deps: [{ token: 'FS_EXAMPLE_CONFIG' }, { token: i1.FsIFrame }], target: i0.ɵɵFactoryTarget.NgModule });
FsExampleModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsExampleModule, declarations: [FsExampleComponent,
        FsExamplesComponent,
        FsExampleHighlightComponent], imports: [CommonModule,
        MatIconModule,
        MatToolbarModule,
        MatTabsModule,
        MatCardModule,
        MatButtonModule,
        MatTooltipModule,
        HttpClientModule, i1.FsIFrameModule, FsDrawerModule,
        HighlightModule], exports: [FsExampleComponent,
        FsExamplesComponent,
        FsExampleHighlightComponent] });
FsExampleModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsExampleModule, providers: [], imports: [[
            CommonModule,
            MatIconModule,
            MatToolbarModule,
            MatTabsModule,
            MatCardModule,
            MatButtonModule,
            MatTooltipModule,
            HttpClientModule,
            FsIFrameModule.forRoot(),
            FsDrawerModule,
            HighlightModule,
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsExampleModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        MatIconModule,
                        MatToolbarModule,
                        MatTabsModule,
                        MatCardModule,
                        MatButtonModule,
                        MatTooltipModule,
                        HttpClientModule,
                        FsIFrameModule.forRoot(),
                        FsDrawerModule,
                        HighlightModule,
                    ],
                    exports: [
                        FsExampleComponent,
                        FsExamplesComponent,
                        FsExampleHighlightComponent,
                    ],
                    declarations: [
                        FsExampleComponent,
                        FsExamplesComponent,
                        FsExampleHighlightComponent,
                    ],
                    providers: [],
                }]
        }], ctorParameters: function () { return [{ type: undefined, decorators: [{
                    type: Inject,
                    args: ['FS_EXAMPLE_CONFIG']
                }] }, { type: i1.FsIFrame }]; } });
