// @firestitch/drawer-unique-namespace/esm5/app/components/drawer-side/drawer-side.component.js
import { __decorate, __metadata } from "tslib";
import { AfterViewInit, ChangeDetectionStrategy, ChangeDetectorRef, Component, ContentChildren, ElementRef, EventEmitter, HostBinding, Input, OnDestroy, OnInit, QueryList, Renderer2, TemplateRef, } from '@angular/core';
import { takeUntil } from 'rxjs/operators';
import { DrawerRef } from '../../classes/drawer-ref';
import { FsDrawerActionDirective } from '../../directives/drawer-action.directive';
var FsDrawerSideComponent = /** @class */ (function () {
    function FsDrawerSideComponent(_cdRef, _renderer, _elRef) {
        this._cdRef = _cdRef;
        this._renderer = _renderer;
        this._elRef = _elRef;
        this.classSide = true;
        this.activeTemplate = null;
        this._hidden = false;
        this._destroy$ = new EventEmitter();
    }
    Object.defineProperty(FsDrawerSideComponent.prototype, "hidden", {
        set: function (value) {
            this._hidden = value;
            if (this._hidden) {
                this._renderer.setAttribute(this._elRef.nativeElement, 'hidden', 'true');
            }
            else {
                this._renderer.removeAttribute(this._elRef.nativeElement, 'hidden');
            }
        },
        enumerable: true,
        configurable: true
    });
    FsDrawerSideComponent.prototype.ngOnInit = function () {
        this.hidden = true;
        if (!this.drawer) {
            console.error('Drawer reference is null for @Input("fsDrawerSide")');
        }
        this._subscribeToActionChanges();
    };
    FsDrawerSideComponent.prototype.ngAfterViewInit = function () {
        var _this = this;
        setTimeout(function () {
            // Check current side status
            if (_this.drawer) {
                _this.hidden = !_this.drawer.isSideOpen;
            }
            _this._updateActiveActionTemplate();
        });
    };
    FsDrawerSideComponent.prototype.ngOnDestroy = function () {
        this._destroy$.emit();
        this._destroy$.complete();
    };
    FsDrawerSideComponent.prototype._subscribeToActionChanges = function () {
        var _this = this;
        if (this.drawer) {
            this.drawer.sideToggle$
                .pipe(takeUntil(this._destroy$))
                .subscribe(function () {
                _this.hidden = !_this.drawer.isSideOpen;
                _this._updateActiveActionTemplate();
                _this._cdRef.detectChanges();
            });
        }
    };
    FsDrawerSideComponent.prototype._updateActiveActionTemplate = function () {
        if (this.drawer) {
            var activatedAction_1 = this.drawer.activeAction;
            if (this.drawer.isSideOpen && activatedAction_1) {
                var selectedActionIndex = this.actions
                    .toArray()
                    .findIndex(function (action) { return action.name === activatedAction_1; });
                this.activeTemplate = this.actionsTemplates.toArray()[selectedActionIndex];
            }
            else {
                this.activeTemplate = null;
            }
        }
    };
    FsDrawerSideComponent.ctorParameters = function () { return [
        { type: ChangeDetectorRef },
        { type: Renderer2 },
        { type: ElementRef }
    ]; };
    __decorate([
        Input('fsDrawerSide'),
        __metadata("design:type", DrawerRef)
    ], FsDrawerSideComponent.prototype, "drawer", void 0);
    __decorate([
        HostBinding('class.side'),
        __metadata("design:type", Object)
    ], FsDrawerSideComponent.prototype, "classSide", void 0);
    __decorate([
        ContentChildren(FsDrawerActionDirective),
        __metadata("design:type", QueryList)
    ], FsDrawerSideComponent.prototype, "actions", void 0);
    __decorate([
        ContentChildren(FsDrawerActionDirective, { read: TemplateRef }),
        __metadata("design:type", QueryList)
    ], FsDrawerSideComponent.prototype, "actionsTemplates", void 0);
    FsDrawerSideComponent = __decorate([
        Component({
            selector: '[fsDrawerSide]',
            template: "<ng-container [ngSwitch]=\"drawer.drawerConfig.position\" *ngIf=\"activeTemplate\">\n  <ng-container *ngSwitchCase=\"'right'\">\n    <div class=\"side-container\" #sideContainer>\n      <div class=\"side-content\">\n        <ng-template [ngTemplateOutlet]=\"activeTemplate\"></ng-template>\n      </div>\n    </div>\n    <div class=\"resize-side\"\n         resizable=\"true\"\n         direction=\"left\"\n         type=\"side\"\n         [fsDrawerResizer]=\"sideContainer\"\n         [resizable]=\"drawer.drawerConfig.resizable\"\n         [sizeController]=\"drawer.resizeController\"\n         [parentContainer]=\"drawer.drawerContentContainer\"\n         [actionsContainer]=\"drawer.drawerActionsContainer\"\n    >\n      <a mat-icon-button class=\"resize-icon\"><mat-icon>drag_indicator</mat-icon></a>\n    </div>\n  </ng-container>\n  <ng-container *ngSwitchCase=\"'left'\">\n    <div class=\"resize-side\"\n         resizable=\"true\"\n         direction=\"right\"\n         [fsDrawerResizer]=\"sideContainer\"\n    >\n    <a mat-icon-button class=\"resize-icon\"><mat-icon>drag_indicator</mat-icon></a>\n    </div>\n    <div class=\"side-container\" #sideContainer>\n      <div class=\"side-content\">\n        <ng-template [ngTemplateOutlet]=\"activeTemplate\"></ng-template>\n      </div>\n    </div>\n  </ng-container>\n</ng-container>\n",
            changeDetection: ChangeDetectionStrategy.OnPush,
            styles: [".resize-icon{width:24px;height:24px;line-height:24px}.resize-icon span{width:24px}"]
        }),
        __metadata("design:paramtypes", [ChangeDetectorRef,
            Renderer2,
            ElementRef])
    ], FsDrawerSideComponent);
    return FsDrawerSideComponent;
}());
export { FsDrawerSideComponent };
