// @firestitch/store/esm2015/app/fs-store.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FsStore } from './services/store.service';
import * as i0 from "@angular/core";
export class FsStoreModule {
    static forRoot() {
        return {
            ngModule: FsStoreModule,
            providers: [
                FsStore
            ],
        };
    }
}
FsStoreModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsStoreModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
FsStoreModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsStoreModule, imports: [CommonModule] });
FsStoreModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsStoreModule, imports: [[
            CommonModule,
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsStoreModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                    ]
                }]
        }] });
