// @firestitch/drawer-unique-namespace/esm5/app/components/drawer-actions/drawer-actions.component.js
import { __decorate, __metadata } from "tslib";
import { ChangeDetectionStrategy, ChangeDetectorRef, Component, Input, OnDestroy, OnInit, } from '@angular/core';
import { Subject } from 'rxjs';
import { FsDrawerAction } from '../../helpers/action-type.enum';
import { DrawerRef } from '../../classes/drawer-ref';
import { FsDrawerMenuService } from '../../services/drawer-menu.service';
var FsDrawerActionsComponent = /** @class */ (function () {
    function FsDrawerActionsComponent(_drawerRef, _drawerMenu, _cdRef) {
        this._drawerRef = _drawerRef;
        this._drawerMenu = _drawerMenu;
        this._cdRef = _cdRef;
        this._destroy$ = new Subject();
    }
    FsDrawerActionsComponent.prototype.ngOnInit = function () {
    };
    FsDrawerActionsComponent.prototype.ngOnDestroy = function () {
        this._destroy$.next();
        this._destroy$.complete();
    };
    FsDrawerActionsComponent.prototype.actionClick = function (action, event) {
        var hasComponentType = action.type === FsDrawerAction.Component;
        var hasMenuType = action.type === FsDrawerAction.Menu;
        if (hasComponentType) {
            var menuRef = this._drawerMenu.create(action.component, event.srcElement, action);
            this._drawerRef.addMenuRef(action.menuRefName, menuRef);
            var params = {
                event: event,
                action: action,
                drawerRef: this._drawerRef,
                menuRef: menuRef
            };
            // Call click
            action.click.call(null, params);
            if (action.closeSide) {
                this._drawerRef.closeSide();
            }
        }
        else if (action.click) {
            var params = { event: event, action: action };
            action.click.call(null, params);
        }
        if (action.close) {
            this._drawerRef.close();
        }
        if (action.toggle && ((!hasComponentType && !hasMenuType) || action.closeSide)) {
            if (this._drawerRef.isSideOpen && this._drawerRef.activeAction === action.name) {
                this._drawerRef.toggleSide();
            }
            else {
                this._drawerRef.setActiveAction(action.name);
            }
        }
    };
    FsDrawerActionsComponent.ctorParameters = function () { return [
        { type: DrawerRef },
        { type: FsDrawerMenuService },
        { type: ChangeDetectorRef }
    ]; };
    __decorate([
        Input(),
        __metadata("design:type", Array)
    ], FsDrawerActionsComponent.prototype, "actions", void 0);
    __decorate([
        Input(),
        __metadata("design:type", String)
    ], FsDrawerActionsComponent.prototype, "activeAction", void 0);
    FsDrawerActionsComponent = __decorate([
        Component({
            selector: 'fs-drawer-actions',
            template: "<div class=\"actions-container\">\n  <div *ngFor=\"let action of actions\" class=\"action-btn\">\n    <fs-drawer-action-item [action]=\"action\"\n                           [activeAction]=\"activeAction\"\n                           (actionClicked)=\"actionClick(action, $event)\">\n    </fs-drawer-action-item>\n  </div>\n</div>\n\n<div class=\"drag-icon\">\n  <a mat-icon-button class=\"resize-icon\"><mat-icon>drag_indicator</mat-icon></a>\n</div>\n",
            changeDetection: ChangeDetectionStrategy.OnPush,
            styles: [":host{display:flex}.actions-container{flex:0}.drag-icon{display:flex;flex:1;align-items:center}"]
        }),
        __metadata("design:paramtypes", [DrawerRef,
            FsDrawerMenuService,
            ChangeDetectorRef])
    ], FsDrawerActionsComponent);
    return FsDrawerActionsComponent;
}());
export { FsDrawerActionsComponent };
