// @firestitch/menu/esm2015/app/components/fs-menu/menu-items-list/menu-items-list.component.js
import { ChangeDetectionStrategy, ChangeDetectorRef, Component, Input, } from '@angular/core';
import { Subject } from 'rxjs';
import { takeUntil } from 'rxjs/operators';
import { createItemsObserver } from '../../../helpers/create-items-observer';
import * as i0 from "@angular/core";
import * as i1 from "@angular/material/menu";
import * as i2 from "@angular/common";
import * as i3 from "@angular/router";
export class MenuItemsListComponent {
    constructor(_cdRef) {
        this._cdRef = _cdRef;
        this._destroy$ = new Subject();
    }
    ngOnChanges(changes) {
        if (changes.items) {
            this._cdRef.detectChanges();
            this._destroy$.next();
            this.subscribeToChanges();
        }
    }
    ngOnDestroy() {
        this._destroy$.next();
        this._destroy$.complete();
    }
    /**
     * For improve ngFor perf
     * @param index
     */
    trackBy(index) {
        return index;
    }
    /**
     * Subscribe to changes in directive parameters.
     * For example we must start detect changes if [hidden] param was changed
     */
    subscribeToChanges() {
        if (this.items && this.items.length) {
            createItemsObserver(this.items)
                .pipe(takeUntil(this._destroy$))
                .subscribe(() => {
                this._cdRef.detectChanges();
            });
        }
    }
}
/** @nocollapse */ MenuItemsListComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.13", ngImport: i0, type: MenuItemsListComponent, deps: [{ token: i0.ChangeDetectorRef }], target: i0.ɵɵFactoryTarget.Component });
/** @nocollapse */ MenuItemsListComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.13", type: MenuItemsListComponent, selector: "fs-menu-items-list", inputs: { items: "items", parentVisible: "parentVisible" }, usesOnChanges: true, ngImport: i0, template: "<ng-container *ngFor=\"let item of items; trackBy: trackBy\">\n  <ng-container *ngIf=\"item.isGroup; else menuItem\">\n    <div class=\"fs-menu-groupped-list\" [hidden]=\"!item.visible\">\n      <div class=\"mat-menu-item fs-menu-group-label\" disabled>\n        <ng-container *ngIf=\"!!item.label; else tpl\">\n          {{ item.label }}\n        </ng-container>\n        <ng-template #tpl\n                     [ngTemplateOutlet]=\"item.groupItemTemplateRef\">\n        </ng-template>\n      </div>\n      <fs-menu-items-list [items]=\"item.childrenItems\"\n                          [parentVisible]=\"item.visible\">\n      </fs-menu-items-list>\n    </div>\n  </ng-container>\n  <ng-template #menuItem>\n    <ng-container *ngIf=\"item.link; else btn\">\n      <a\n        mat-menu-item\n        *ngIf=\"item.visible\"\n        [routerLink]=\"item.link\"\n        [queryParams]=\"item.queryParams\"\n        [target]=\"item.target\"\n        (click)=\"item.click($event)\"\n        [class]=\"item.cssClass\"\n        [ngClass]=\"item.ngClass\"\n        [id]=\"item.cssId\">\n        <ng-template [ngTemplateOutlet]=\"item.templateRef\"></ng-template>\n      </a>\n    </ng-container>\n    <ng-template #btn>\n      <button type=\"button\"\n              mat-menu-item\n              *ngIf=\"item.visible\"\n              (click)=\"item.click($event)\"\n              [class]=\"item.cssClass\"\n              [ngClass]=\"item.ngClass\"\n              [id]=\"item.cssId\"\n      >\n        <ng-template [ngTemplateOutlet]=\"item.templateRef\"></ng-template>\n      </button>\n    </ng-template>\n  </ng-template>\n</ng-container>\n", styles: [":host ::ng-deep .fs-menu-groupped-list a,:host ::ng-deep .fs-menu-groupped-list button{padding-left:32px}\n"], components: [{ type: MenuItemsListComponent, selector: "fs-menu-items-list", inputs: ["items", "parentVisible"] }, { type: i1.MatMenuItem, selector: "[mat-menu-item]", inputs: ["disabled", "disableRipple", "role"], exportAs: ["matMenuItem"] }], directives: [{ type: i2.NgForOf, selector: "[ngFor][ngForOf]", inputs: ["ngForOf", "ngForTrackBy", "ngForTemplate"] }, { type: i2.NgIf, selector: "[ngIf]", inputs: ["ngIf", "ngIfThen", "ngIfElse"] }, { type: i2.NgTemplateOutlet, selector: "[ngTemplateOutlet]", inputs: ["ngTemplateOutletContext", "ngTemplateOutlet"] }, { type: i3.RouterLinkWithHref, selector: "a[routerLink],area[routerLink]", inputs: ["routerLink", "target", "queryParams", "fragment", "queryParamsHandling", "preserveFragment", "skipLocationChange", "replaceUrl", "state", "relativeTo"] }, { type: i2.NgClass, selector: "[ngClass]", inputs: ["class", "ngClass"] }], changeDetection: i0.ChangeDetectionStrategy.OnPush });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.13", ngImport: i0, type: MenuItemsListComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'fs-menu-items-list',
                    templateUrl: './menu-items-list.component.html',
                    styleUrls: ['./menu-items-list.component.scss'],
                    changeDetection: ChangeDetectionStrategy.OnPush,
                }]
        }], ctorParameters: function () { return [{ type: i0.ChangeDetectorRef }]; }, propDecorators: { items: [{
                type: Input
            }], parentVisible: [{
                type: Input
            }] } });
