// @firestitch/message/esm2015/app/components/messages/messages.component.js
import { ChangeDetectionStrategy, ChangeDetectorRef, Component } from '@angular/core';
import { Subject } from 'rxjs';
import { takeUntil } from 'rxjs/operators';
import { remove } from 'lodash-es';
import { FsMessage } from '../../message.service';
import * as i0 from "@angular/core";
import * as i1 from "../../message.service";
import * as i2 from "../message/message.component";
import * as i3 from "@angular/common";
export class FsMessagesComponent {
    constructor(_fsMessage, _cdRef) {
        this._fsMessage = _fsMessage;
        this._cdRef = _cdRef;
        this.messages = [];
        this._destroy$ = new Subject();
    }
    ngOnInit() {
        this._fsMessage
            .bannerMessages$
            .pipe(takeUntil(this._destroy$))
            .subscribe((message) => {
            if (!message) {
                this.messages = [];
            }
            else {
                this.messages.push(message);
                if (message.timeout) {
                    setTimeout(() => {
                        remove(this.messages, (item) => {
                            return item === message;
                        });
                        this._cdRef.markForCheck();
                    }, message.timeout);
                }
            }
            this._cdRef.markForCheck();
        });
    }
    ngOnDestroy() {
        this._destroy$.next();
        this._destroy$.complete();
    }
}
FsMessagesComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsMessagesComponent, deps: [{ token: i1.FsMessage }, { token: i0.ChangeDetectorRef }], target: i0.ɵɵFactoryTarget.Component });
FsMessagesComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.16", type: FsMessagesComponent, selector: "fs-messages", ngImport: i0, template: `<fs-message *ngFor="let item of messages" [fsType]="item.type" [fsMessage]="item.msg"></fs-message>`, isInline: true, components: [{ type: i2.FsMessageComponent, selector: "fs-message", inputs: ["type", "message", "fsType", "fsMessage"] }], directives: [{ type: i3.NgForOf, selector: "[ngFor][ngForOf]", inputs: ["ngForOf", "ngForTrackBy", "ngForTemplate"] }], changeDetection: i0.ChangeDetectionStrategy.OnPush });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FsMessagesComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'fs-messages',
                    template: `<fs-message *ngFor="let item of messages" [fsType]="item.type" [fsMessage]="item.msg"></fs-message>`,
                    changeDetection: ChangeDetectionStrategy.OnPush,
                }]
        }], ctorParameters: function () { return [{ type: i1.FsMessage }, { type: i0.ChangeDetectorRef }]; } });
